### ISSUE TITLE
deadlock detected while deleting layer

### ISSUE THREAD

**Issue by juanignaciosl** [CartoDB/cartodb#11443]
```ruby
PG::Error: ERROR: deadlock detected
DETAIL: Process 30804 waits for ShareLock on transaction 473226780; blocked by process 32674.
Process 32674 waits for ShareLock on transaction 473226783; blocked by process 30804.
HINT: See server log for query details.

+ 2 non-project frames
3
File "/home/ubuntu/www/production.cartodb.com/releases/20170202144204/config/initializers/zz_patch_reconnect.rb", line 51 in block in exec_delete
4
File "/home/ubuntu/www/production.cartodb.com/releases/20170202144204/config/initializers/zz_patch_reconnect.rb", line 62 in block (2 levels) in wrap_execute
+ 3 non-project frames
8
File "/home/ubuntu/www/production.cartodb.com/releases/20170202144204/config/initializers/zz_patch_reconnect.rb", line 61 in block in wrap_execute
9
File "/home/ubuntu/www/production.cartodb.com/releases/20170202144204/config/initializers/zz_patch_reconnect.rb", line 68 in with_auto_reconnect
10
File "/home/ubuntu/www/production.cartodb.com/releases/20170202144204/config/initializers/zz_patch_reconnect.rb", line 60 in wrap_execute
11
File "/home/ubuntu/www/production.cartodb.com/releases/20170202144204/config/initializers/zz_patch_reconnect.rb", line 50 in exec_delete
+ 17 non-project frames
29
File "/home/ubuntu/www/production.cartodb.com/releases/20170202144204/app/controllers/carto/api/layers_controller.rb", line 160 in destroy
30
File "/home/ubuntu/www/production.cartodb.com/releases/20170202144204/app/controllers/carto/api/layers_controller.rb", line 70 in map_destroy
+ 6 non-project frames
37
File "/home/ubuntu/www/production.cartodb.com/releases/20170202144204/app/controllers/application_controller.rb", line 100 in wrap_in_profiler
+ 73 non-project frames

```

https://rollbar.com/vizzuality/CartoDB/items/21610/?item_page=0&#instances

**Comment by juanignaciosl** [comment]
Closing automatically because of inactivity. If anybody considers that this still important, please reopen and it'll be prioritized.

**Comment by juanignaciosl** [comment]
@CartoDB/builder-backend it'd be great if next refactor step could address this bug.

**Comment by javitonino** [comment]
Little research from last time this happened. It seems this is not entirely related with multiple ORMs, as we updated the layers controller to use only new models precisely for this but it kept happening.

Take a look at:
https://rollbar.com/vizzuality/CartoDB/items/21612/occurrences/21704486652/
https://rollbar.com/vizzuality/CartoDB/items/21445/occurrences/21704486584/

There are two simultaneous PUTs to the same layer, which is the problem. Maybe we should wrap the controller in a transaction, which should be possible now that everything uses the same model :)

**Comment by juanignaciosl** [comment]
Uhm, there's a loop with multiple DB operations without a transaction, that's not right. And you nailed it with those requests, they're blocking each other ✨ 

At first sight, a transaction should fix it. Nevertheless, if there are other operations that don't use _that_ transaction it could lead to blocks within a single request, or fetching of old data, so this should be reviewed and tested with care.

Good catch!

**Comment by javitonino** [comment]
Ok, finally found a likely explanation for the issue, but it's very hard to be sure. The problem here is not the lack of transaction. Although we should add them, this part of the controller (the loop) is only used for reordering layer, and is not the case in any of the deadlocks I found.

The main problem I get looking at Rollbar, is that you only get N-1 messages where N is the number of processes involved in the deadlock. That is because PostgreSQL kills all deadlocked queries except for one, that succeeds. The server log includes all N queries, but lacks context (ruby stacktrace).

Anyway, this is the case that I managed to reproduce, simultaneous requests to update and delete a layer. The update consists of adding/deleting an analysis node with a JOIN. What I see here is the user adding a JOIN with network problems, it doesn't work, so they delete the layer and retry.

What happens behind the scenes has to do with the `layers` and `layers_user_tables` relation. The update has `register_table_dependencies` in an after hook, while in the deletion, the dependencies must be removed before the layer (for FK consistency).

So what ends up happening is that the transactions run like this:
1. Update controller updates the layer
2. Delete controller removes layers_user_tables
3. Update controller tries to add/remove layers_user_tables. Starts waiting for a lock
4. Delete controller tries to remove the layer. Starts waiting for a lock. Deadlock is inmediately detected and this request is killed (500).
5. Update completes successfully.
